### PR TITLE
DOP-5124: reconstruct table data using tanstack tables

### PR DESCRIPTION
**Pull request by seungpark**
### Stories/Links:

DOP-5124

### Current Behavior:

[test page for tables on main branch](https://docs-mongodb-org-stg.s3.us-east-2.amazonaws.com/sandbox/cloud-docs/seung.park/main/test-tables/index.html) - includes all of Atlas docs as well as the test page (sourced [here](https://github.com/rayangler/cloud-docs/blob/sandbox/source/test-tables.txt))

### Staging Links:
[test page for tables on branch](https://docs-mongodb-org-stg.s3.us-east-2.amazonaws.com/DOP-5124-test-tables/cloud-docs/seung.park/DOP-5124/test-tables/index.html)

### Notes:
- This PR does not change the way tables are rendered, or how they should be used in ReST. This updates the logic for constructing table data on the front end, so that we can pass processed column/row data, enabling us to use farther features as [expanded content ](https://github.com/mongodb/leafygreen-ui/tree/main/packages/table#data--renderexpandedcontent) and/or nested tables.

### README updates

- - [ ] This PR introduces changes that should be reflected in the README, and I have made those updates.
- - [X] This PR does not introduce changes that should be reflected in the README

### CODE REVIEW

**Review by biancalaube — COMMENTED** [review]
I noticed that in the staging link the `Nested table with header rows` had some weird spacing. I think both the columns in the staged link are the same size, but in the main link the `service` column has a smaller width. Is this expected or do we want it to look as it does originally? 
new link ----
Screenshot 2024-12-03 at 4 27 56 PM
old link ----- 
Screenshot 2024-12-03 at 4 30 45 PM

**Comment by seungpark** [comment]
> I noticed that in the staging link the `Nested table with header rows` had some weird spacing. I think both the columns in the staged link are the same size, but in the main link the `service` column has a smaller width. Is this expected or do we want it to look as it does originally? new link ---- Screenshot 2024-12-03 at 4 27 56 PM old link ----- Screenshot 2024-12-03 at 4 30 45 PM

addressed! [removed the styling](https://github.com/mongodb/snooty/pull/1316/commits/296ab80c1e7cadb09d707e6196e03f9e555438ef) - this was actually put in place to fix an existing issue, but that was resolved by getting the latest commit from cloud docs (seems like it was a source ReST issue)

updated the staging links above

File: src/components/ListTable.js [hunk 1]
```diff
@@ -159,43 +177,69 @@ const includesNestedTable = (rows) => {
...
+  // TODO: header row count should not be more than 1
+  // need a warning in parser
   const headerRowCount = parseInt(options?.['header-rows'], 10) || 0;
   const stubColumnCount = parseInt(options?.['stub-columns'], 10) || 0;
   const bodyRows = children[0].children.slice(headerRowCount);
   const columnCount = bodyRows[0].children[0].children.length;
 
   // Check if :header-rows: 0 is specified or :header-rows: is omitted
-  const headerRows = headerRowCount > 0 ? children[0].children[0].children.slice(0, headerRowCount) : [];
+  const [headerRows] = useState(() =>
```
[the remark this answers is not in the document]
Comment (seungpark) [reply src/components/ListTable.js:240]: i actually initially had used useMemo hooks with dependencies that should not change (derived from nodeData) but noticed this was pretty taxing and causing page freezes. i opted to switch to a state that only initializes and does not update in [this commit](https://github.com/mongodb/snooty/pull/1316/commits/674f0901f5db93a612b936454969d2522b8195a0)

File: src/components/ListTable.js [hunk 2]
```diff
@@ -159,43 +177,69 @@ const includesNestedTable = (rows) => {
...
-      return (
-        <Cell key={colIndex} className={cx(baseCellStyle, bodyCellStyle, isStub && stubCellStyle)} role={role}>
-          {/* Wrap in div to ensure contents are structured properly */}
-          <div>{contents}</div>
-        </Cell>
+const generateColumns = (headerRow, bodyRows) => {
+  if (!headerRow?.children) {
+    // generate columns from bodyRows
+    const flattenedRows = bodyRows.map((bodyRow) => bodyRow.children[0].children);
+    const maxColumns = Math.max(...flattenedRows.map((row) => row.length));
```
[the remark this answers is not in the document]
Comment (seungpark) [reply src/components/ListTable.js:184]: i recall an index error where a row's element couldnt find a column reference in [this line](https://github.com/mongodb/snooty/pull/1316/files#diff-c0635850a29537b5e9333b6e82a9306b70f036a6c42f9c129a903418a00ac4ebR216) so i added this safeguard.

good call on the column count! will use this output of this function to calculate that and consolidate the two

File: src/components/ListTable.js [hunk 3]
```diff
@@ -159,43 +177,69 @@ const includesNestedTable = (rows) => {
...
+  });
+
+  return rows;
 };
 
 const ListTable = ({ nodeData: { children, options }, ...rest }) => {
   const ancestors = useAncestorComponentContext();
-  const { theme: siteTheme } = useDarkMode();
+  // TODO: header row count should not be more than 1
+  // need a warning in parser
```
[the remark this answers is not in the document]
Comment (seungpark) [reply src/components/ListTable.js:233]: done!

File: src/components/ListTable.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (seungpark) [reply src/components/ListTable.js:184]: merged the logic of column count!

File: src/components/ListTable.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (rayangler) [reply src/components/ListTable.js]: Interesting. I wonder if the bad performance was due to `bodyRows` not being memoized? Seems like we're slicing an array, which results in a new array that was potentially causing dependencies to always change?

Would something like:

```js
const bodyRows = useMemo(() => children[0].children.slice(headerRowCount), [children, headerRowCount]);
```

work?

[the remark this answers is not in the document]
Comment (seungpark) [reply src/components/ListTable.js]: good call! done and updated the above link

**Review by rayangler — APPROVED** [review]

**Pull request merged**